clang/lib/Driver/Driver.cpp

```diff
@@ -5134,6 +5136,49 @@ const ToolChain &Driver::getToolChain(const ArgList &Args,
   return *TC;
 }
 
+const ToolChain &Driver::getOffloadingDeviceToolChain(const ArgList &Args,
+                  const llvm::Triple &Target, const ToolChain &HostTC,
+                  const Action::OffloadKind &TargetDeviceOffloadKind) const {
+  // Use device / host triples as the key into the ToolChains map because the
+  // device ToolChain we create depends on both.
+  auto &TC = ToolChains[Target.str() + "/" + HostTC.getTriple().str()];
+  if (!TC) {
+    // Categorized by offload kind > arch rather than OS > arch like
+    // the normal getToolChain call, as it seems a reasonable way to categorize
+    // things.
+    switch (TargetDeviceOffloadKind) {
+      case Action::OFK_Cuda:
+        TC = llvm::make_unique<toolchains::CudaToolChain>(
+          *this, Target, HostTC, Args, TargetDeviceOffloadKind);
+        break;
+      case Action::OFK_HIP:
+        TC = llvm::make_unique<toolchains::HIPToolChain>(
+          *this, Target, HostTC, Args);
+        break;
+      case Action::OFK_OpenMP:
+        // omp + nvptx
+        TC = llvm::make_unique<toolchains::CudaToolChain>(
+          *this, Target, HostTC, Args, TargetDeviceOffloadKind);
+        break;
+      case Action::OFK_SYCL:
+        switch (Target.getArch()) {
```

Comment:
Can we remove this switch now? 
I think we can assume that target is valid. Right?

Reply:
Yes, that would be no problem, I could remove the switch from the commit if you'd like as it serves no purpose at the moment other than indicating the intent is to allow extension via a switch similar to getToolChain?

As in our case it's required and the intent is to be able to swap between other viable SYCL ToolChains based on target: https://github.com/triSYCL/sycl/blob/sycl/unified/master/clang/lib/Driver/Driver.cpp#L5142

Reply:
Thanks for clarification.
I'm fine with leaving the switch then.

Reply:
In reference to the validity check though, the check occurs after this function has been invoked at the moment rather than before. If we wish to preemptively check and never risk the chance of returning an invalid ToolChain the check would have to be moved to here: https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/Driver.cpp#L780 

Which makes sense to me, however, It would be a little different from OpenMP and CUDA as they both seem to check after ToolChain selection. 

In both cases invalid targets should be caught though, just currently it's later rather than sooner.